doc/config-replace/SONiC_Configuration_Replace_Design_Doc.md

```diff
@@ -0,0 +1,537 @@
+- 1.3.0 - User should be able to preview  the configuration difference before going ahead and affecting the configuration changes
+- 1.4.0 - A reliable fallback mechanism must be implemented to recover from any errors encountered during the configuration replace operation
+
+
+
+### 1.1.2 Configuration and Management Requirements
+- 2.1.0 - Python-click based CLI command *config replace* must be implemented to perform the SONiC configuration replace operation
+
+- 2.2.0 - *config replace* equivalent KLISH CLI command should be implemented
+
+- 2.3.0 - Only root user must be allowed to execute the *config replace* command
+
+- 2.4.0 - The *config replace* command must provide following sub options:
+  - 2.4.0.1 - "show-patch-" Perform a dry run of the operation to validate if the config replace action can be executed and display the configuration difference that would be applied
+  
+  - 2.4.0.2 - verbose - Provide additional information on the configuration steps being performed
+  
+  - 2.4.0.3 - debug - Print details on each individual ConfigDB update command executed as part of the *config replace* command
+  
+  - 2.4.0.4 - best-effort - Continue with the config replace command even when an error is encountered. Only operational errors are ignored, any command or input syntax errors are flagged and the command exits.
+  
+  - 2.4.0.5 - wait-time - Maximum duration in seconds for which the *config replace* command waits to confirm that the configuration updates for a ConfigDB table have been completely processed by the application before proceeding to the next ConfigDB table. A default value of 30s is used.
+
+
+
+### 1.1.3 Scalability Requirements
+
+N/A
+
+### 1.1.4 Warm Boot Requirements
+
+N/A
+
+## 1.2 Design Overview
+### 1.2.1 Basic Approach
+When the SONiC ConfigDB contents are stored in a file format, they are stored in JSON format. Even the user specifies the intended configuration state as a JSON file. Hence the difference between two SONiC configuration states can be described as the difference between two JSON files. The notation of [RFC 6902 JSON Patch](https://tools.ietf.org/html/rfc6902) is used to represent this configuration difference.
+
+
+
+The [Jsonpatch](https://pypi.org/project/jsonpatch/) python library is used to calculate the RFC6902 compliant JSON patch file between the running configuration and user specified *config_db.json*.
+
+
+
+The  *ConfigDBConnector* class of the *sonic-py-swsssdk* is used to obtain the running configuration in JSON format. A new python script *config_replace* is introduced which calculates the JSON patch and then translates the JSON patch to equivalent Redis DB commands to update the ConfigDB. The *ConfigDBConnetor* class is used by the *config_replace* script to perform DB updates in ConfigDB instance of the redis database.
```

Comment:
SONiC configLet does similar job of JSON Patch and it uses swssdk, please evaluate and try to leverage as much as possible. please refer- https://github.com/Azure/sonic-utilities/blob/master/scripts/configlet.